File: Data Structures & Algorithms/max-area-of-island/submission-1.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        ROWS, COLS = len(grid), len(grid[0])
        res = 0
        
        def bfs(r, c):
            q = deque()
            q.append([r, c])
            grid[r][c] = 0
            length = 0
            while q:
                row, col = q.popleft()
                length += 1
                for dr, dc in directions:
                    newRow = row + dr
                    newCol = col + dc
                    if min(newRow, newCol) < 0 or newRow >= ROWS or newCol >= COLS or grid[newRow][newCol] == 0:
                        continue
                    q.append([newRow, newCol])
                    grid[newRow][newCol] = 0
            return length
        
        for row in range(ROWS):
            for col in range(COLS):
                if grid[row][col] == 1:
                    res = max(res, bfs(row, col))
        return res

File: Data Structures & Algorithms/count-number-of-islands/submission-1.java
class Solution {
    int ROWS;
    int COLS;
    public int numIslands(char[][] grid) {
        ROWS = grid.length;
        COLS = grid[0].length;

        int count = 0;
        for(int i=0; i<ROWS; ++i) {
            for(int j=0; j<COLS; ++j) {
                if(grid[i][j] == '1') {
                    count++;
                    dfs(i, j, grid);
                }
            }
        }
        return count;
    }

    private void dfs(int r, int c, char[][] grid) {
        if(Math.min(r, c) < 0 || r >= ROWS || c >= COLS || grid[r][c] == '0') {
            return;
        }
        grid[r][c] = '0';
        dfs(r + 1, c, grid);
        dfs(r - 1, c, grid);
        dfs(r, c + 1, grid);
        dfs(r, c - 1, grid);
    }
}
